ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecryptPioneerPayloads.java

```diff
@@ -0,0 +1,87 @@
+import java.util.Collections;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.FlatMapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ProcessFunction;
+import org.apache.beam.sdk.transforms.WithFailures;
+import org.apache.beam.sdk.transforms.WithFailures.Result;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.jose4j.jwe.JsonWebEncryption;
+import org.jose4j.lang.JoseException;
+
+public class DecryptPioneerPayloads extends
+    PTransform<PCollection<PubsubMessage>, Result<PCollection<PubsubMessage>, PubsubMessage>> {
+
+  private final ValueProvider<String> metadataLocation;
+  private final ValueProvider<Boolean> kmsEnabled;
+  private transient KeyStore keyStore;
+
+  public static DecryptPioneerPayloads of(ValueProvider<String> metadataLocation,
+      ValueProvider<Boolean> kmsEnabled) {
+    return new DecryptPioneerPayloads(metadataLocation, kmsEnabled);
+  }
+
+  private DecryptPioneerPayloads(ValueProvider<String> metadataLocation,
+      ValueProvider<Boolean> kmsEnabled) {
+    this.metadataLocation = metadataLocation;
+    this.kmsEnabled = kmsEnabled;
+  }
+
+  @Override
+  public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
+      PCollection<PubsubMessage> messages) {
+    return messages.apply(FlatMapElements.into(TypeDescriptor.of(PubsubMessage.class)) //
+        .via(new Fn()) //
+        .exceptionsInto(TypeDescriptor.of(PubsubMessage.class)) //
+        .exceptionsVia((WithFailures.ExceptionElement<PubsubMessage> ee) -> {
+          try {
+            throw ee.exception();
+          } catch (IOException e) {
+            return FailureMessage.of(DecryptPioneerPayloads.class.getSimpleName(), //
+                ee.element(), //
+                ee.exception());
+          }
+        }));
+  }
+
+  /**
+   * Decrypt a payload encoded in a compact serialization of JSON Web Encryption (JWE).
+   */
+  public static byte[] decrypt(PrivateKey key, String payload) throws JoseException {
+    JsonWebEncryption jwe = new JsonWebEncryption();
+    jwe.setKey(key);
+    jwe.setContentEncryptionKey(key.getEncoded());
+    jwe.setCompactSerialization(payload);
+    return jwe.getPlaintextBytes();
+  }
+
+  private class Fn implements ProcessFunction<PubsubMessage, Iterable<PubsubMessage>> {
+
+    @Override
+    public Iterable<PubsubMessage> apply(PubsubMessage message) throws IOException, JoseException {
+      message = PubsubConstraints.ensureNonNull(message);
+
+      if (keyStore == null) {
+        keyStore = KeyStore.of(metadataLocation.get(), kmsEnabled.get());
+      }
+
+      ObjectNode json = Json.readObjectNode(message.getPayload());
+      PrivateKey key = keyStore.getKey(message.getAttribute(Attribute.DOCUMENT_NAMESPACE));
+
+      byte[] decrypted = decrypt(key, json.get("payload").asText());
```

Comment:
```suggestion
      byte[] decrypted = decrypt(key, json.get(FieldName.PAYLOAD).asText());
```

Reply:
Do you know what module this comes from?

Reply:
`com.mozilla.telemetry.ingestion.core.Constant.FieldName`